File: 599.MinimumIndexSumOfTwoLists/minIndex.cc
class Solution {
public:
    vector<string> findRestaurant(vector<string>& list1, vector<string>& list2) {
        map<string, int> counts;
        for (int i = 0; i < int(list1.size()); i++) {
            counts[list1[i]] = i;
        }

        vector<string> ans;
        int minIndex = list1.size() + list2.size();
        for (int i = 0; i < int(list2.size()); i++) {
            string v = list2[i];
            if (counts.find(v) == counts.end()) {
                continue;
            }
            int sumIndex = counts[v] + i;
            if (sumIndex < minIndex) {
                minIndex = sumIndex;
                ans.clear();
                ans.push_back(v);
            }
            else if (sumIndex == minIndex) {
                ans.push_back(v);
            }
        }

        return ans;
    }
};

File: 599.MinimumIndexSumOfTwoLists/minIndex.py
class Solution(object):
    def findRestaurant(self, list1, list2):
        """
        :type list1: List[str]
        :type list2: List[str]
        :rtype: List[str]
        """
        counts = {}
        for i in range(len(list1)):
            counts[list1[i]] = i

        ans = []
        minIndex = len(list1) + len(list2)
        for i in range(len(list2)):
            v = list2[i]
            if v not in counts:
                continue
            sumIndex = counts[v] + i
            if sumIndex < minIndex:
                minIndex = sumIndex
                ans = []
                ans.append(v)
            elif sumIndex == minIndex:
                ans.append(v)

        return ans

if __name__ == '__main__':
    list1 = ["Shogun","Tapioca Express","Burger King","KFC"]
    list2 = ["KFC","Shogun","Burger King"]
    print Solution().findRestaurant(list1, list2)
